torch/lib/c10d/ProcessGroupMPI.cpp

```diff
@@ -293,4 +309,200 @@ std::shared_ptr<ProcessGroup::Work> ProcessGroupMPI::allreduce(
   return enqueue(std::move(entry));
 }
 
+std::shared_ptr<ProcessGroup::Work> ProcessGroupMPI::reduce(
+    std::vector<at::Tensor>& tensors,
+    const ReduceOptions& opts) {
+  checkSingleTensor(tensors);
+
+  std::function<void(std::unique_ptr<WorkEntry>&)> runFunc =
+      [opts, this](std::unique_ptr<WorkEntry>& entry) {
+        auto data = (*entry->src)[0];
+        auto dataPtr = (*entry->src)[0].data_ptr();
+        void* sendbuf = (rank_ == opts.rootRank) ? MPI_IN_PLACE : dataPtr;
+        void* recvbuf = (rank_ == opts.rootRank) ? dataPtr : nullptr;
+
+        MPI_CHECK(MPI_Reduce(
+            sendbuf,
+            recvbuf,
+            data.numel(),
+            mpiDatatype.at(data.type().scalarType()),
+            mpiOp.at(opts.reduceOp),
+            opts.rootRank,
+            MPI_COMM_WORLD));
+      };
+  auto entry = std::unique_ptr<WorkEntry>(
+      new WorkEntry(&tensors, nullptr, std::move(runFunc)));
+  return enqueue(std::move(entry));
+}
+
+std::shared_ptr<ProcessGroup::Work> ProcessGroupMPI::allgather(
+    std::vector<std::vector<at::Tensor>>& outputTensors,
+    std::vector<at::Tensor>& inputTensors) {
+  checkSingleTensor(inputTensors);
+  if (outputTensors.size() != 1) {
+    throw std::runtime_error(
+        "MPI process group only supports a single "
+        "tensor op");
```

Comment:
This error message doesn't really explain what's going on. What's a "single tensor op"?

Reply:
addressed